File: benchmarks/julia/benchmark_scaling.jl
#!/usr/bin/env julia
"""
Scaling benchmark: Test Julia FFT-CN performance across grid sizes.
Equivalent to Python benchmark_scaling.py.

Tests: 64², 128², 256², 512², 1024²
"""

using Printf
using Statistics
using JSON
using FFTW
using LinearAlgebra

# Check for CUDA availability
const HAS_CUDA = try
    using CUDA
    CUDA.functional()
catch
    false
end

if HAS_CUDA
    using CUDA
    using CUDA.CUFFT
end

println("=" ^ 60)
println("Julia FFT-CN Scaling Benchmark")
println("=" ^ 60)
println()

# =============================================================================
# Solver Functions (same as benchmark_cpu_gpu.jl)
# =============================================================================

function fftfreq(n::Int, d::Float64=1.0)
    results = zeros(n)
    N = (n - 1) ÷ 2 + 1
    for i in 0:(N-1)
        results[i+1] = i / (n * d)
    end
    for i in N:(n-1)
        results[i+1] = (i - n) / (n * d)
    end
    return results
end

function create_laplacian_eigenvalues(N::Int, dx::Float64)
    kx = fftfreq(N, dx) .* 2π
    ky = fftfreq(N, dx) .* 2π
    KX = repeat(kx, 1, N)
    KY = repeat(ky', N, 1)
    return -(KX.^2 .+ KY.^2)
end

function solve_diffusion_cpu(u0::Matrix{Float64}, D::Float64, dt::Float64, n_steps::Int)
    N = size(u0, 1)
    dx = 1.0 / N

    lap_eig = create_laplacian_eigenvalues(N, dx)
    cn_factor = (1.0 .+ 0.5 .* dt .* D .* lap_eig) ./ (1.0 .- 0.5 .* dt .* D .* lap_eig)

    u = copy(u0)
    u_hat = zeros(ComplexF64, N, N)

    plan_fft = FFTW.plan_fft!(u_hat)
    plan_ifft = FFTW.plan_ifft!(u_hat)

    for _ in 1:n_steps
        u_hat .= u
        plan_fft * u_hat
        u_hat .*= cn_factor
        plan_ifft * u_hat
        u .= real.(u_hat)
    end

    return u
end

if HAS_CUDA

function create_laplacian_eigenvalues_gpu(N::Int, dx::Float64)
    kx = fftfreq(N, dx) .* 2π
    ky = fftfreq(N, dx) .* 2π
    KX = repeat(kx, 1, N)
    KY = repeat(ky', N, 1)
    lap_eig = -(KX.^2 .+ KY.^2)
    return CuArray(lap_eig)
end

function solve_diffusion_gpu(u0::Matrix{Float64}, D::Float64, dt::Float64, n_steps::Int)
    N = size(u0, 1)
    dx = 1.0 / N

    lap_eig = create_laplacian_eigenvalues_gpu(N, dx)
    cn_factor = CuArray(ComplexF64.((1.0 .+ 0.5 .* dt .* D .* Array(lap_eig)) ./ (1.0 .- 0.5 .* dt .* D .* Array(lap_eig))))

    # Work in complex space, use in-place FFT plans
    u = CuArray(ComplexF64.(u0))

    # Create in-place CUFFT plans (much faster than allocating each call)
    plan_fft = CUFFT.plan_fft!(u)
    plan_ifft = CUFFT.plan_ifft!(u)

    for _ in 1:n_steps
        plan_fft * u      # In-place forward FFT
        u .*= cn_factor   # Apply CN factor
        plan_ifft * u     # In-place inverse FFT
    end

    CUDA.synchronize()
    return Array(real.(u))
end

end # HAS_CUDA

# =============================================================================
# Benchmark Configuration
# =============================================================================

const GRID_SIZES = [64, 128, 256, 512, 1024]
const D = 0.01  # Match JAX benchmark
const N_STEPS = 1000  # Match JAX benchmark (1000 steps)
const N_WARMUP = 2
const N_RUNS = 5

println("Configuration:")
println("  Grid sizes: ", GRID_SIZES)
println("  Diffusion coefficient: $D")
println("  Steps per run: $N_STEPS")
println("  Warmup runs: $N_WARMUP")
println("  Benchmark runs: $N_RUNS")
println()

# =============================================================================
# Run Scaling Tests
# =============================================================================

results = Dict{Int, Dict{String, Any}}()

for N in GRID_SIZES
    println("-" ^ 40)
    println("Grid size: $N × $N ($(N^2) DOFs)")
    println("-" ^ 40)

    dx = 1.0 / N
    dt = 0.1 * dx^2 / D  # CFL-like condition

    # Create initial condition
    x = range(0, 1, length=N+1)[1:end-1]
    y = range(0, 1, length=N+1)[1:end-1]
    X = repeat(collect(x), 1, N)
    Y = repeat(collect(y)', N, 1)
    u0 = sin.(2π .* X) .* sin.(2π .* Y)

    # CPU benchmark
    println("  CPU warmup...")
    for _ in 1:N_WARMUP
        solve_diffusion_cpu(u0, D, dt, N_STEPS)
    end

    cpu_times = Float64[]
    for i in 1:N_RUNS
        t = @elapsed solve_diffusion_cpu(u0, D, dt, N_STEPS)
        push!(cpu_times, t)
    end
    cpu_mean = mean(cpu_times)
    @printf("  CPU: %.4f s (%.2f ms/step)\n", cpu_mean, 1000*cpu_mean/N_STEPS)

    result = Dict(
        "N" => N,
        "DOFs" => N^2,
        "dt" => dt,
        "cpu_mean_s" => cpu_mean,
        "cpu_times_s" => cpu_times
    )

    # GPU benchmark
    if HAS_CUDA
        println("  GPU warmup...")
        for _ in 1:N_WARMUP
            solve_diffusion_gpu(u0, D, dt, N_STEPS)
        end

        gpu_times = Float64[]
        for i in 1:N_RUNS
            CUDA.synchronize()
            t = @elapsed begin
                solve_diffusion_gpu(u0, D, dt, N_STEPS)
                CUDA.synchronize()
            end
            push!(gpu_times, t)
        end
        gpu_mean = mean(gpu_times)
        speedup = cpu_mean / gpu_mean
        @printf("  GPU: %.4f s (%.2f ms/step)\n", gpu_mean, 1000*gpu_mean/N_STEPS)
        @printf("  Speedup: %.1fx\n", speedup)

        result["gpu_mean_s"] = gpu_mean
        result["gpu_times_s"] = gpu_times
        result["speedup"] = speedup
    end

    results[N] = result
    println()
end

# =============================================================================
# Summary Table
# =============================================================================

println("=" ^ 60)
println("SCALING SUMMARY")
println("=" ^ 60)
println()

if HAS_CUDA
    @printf("%-8s  %-10s  %-10s  %-10s  %-8s\n", "N", "DOFs", "CPU (s)", "GPU (s)", "Speedup")
    println("-" ^ 50)
    for N in GRID_SIZES
        r = results[N]
        @printf("%-8d  %-10d  %-10.4f  %-10.4f  %-8.1fx\n",
                N, r["DOFs"], r["cpu_mean_s"], r["gpu_mean_s"], r["speedup"])
    end
else
    @printf("%-8s  %-10s  %-10s\n", "N", "DOFs", "CPU (s)")
    println("-" ^ 30)
    for N in GRID_SIZES
        r = results[N]
        @printf("%-8d  %-10d  %-10.4f\n", N, r["DOFs"], r["cpu_mean_s"])
    end
end
println()

# =============================================================================
# Save Results
# =============================================================================

output = Dict(
    "julia_version" => string(VERSION),
    "cuda_available" => HAS_CUDA,
    "config" => Dict(
        "grid_sizes" => GRID_SIZES,
        "D" => D,
        "n_steps" => N_STEPS,
        "n_runs" => N_RUNS
    ),
    "results" => [results[N] for N in GRID_SIZES]
)

if HAS_CUDA
    output["gpu_device"] = CUDA.name(CUDA.device())
end

output_path = joinpath(@__DIR__, "results_scaling.json")
open(output_path, "w") do f
    JSON.print(f, output, 2)
end
println("Results saved to: $output_path")

File: benchmarks/benchmark_scaling.py
#!/usr/bin/env python3
"""
Benchmark: Scaling Study (Table 10)

2D Diffusion with Crank-Nicolson + FFT, 1000 steps
Grid sizes: 64x64, 128x128, 256x256, 512x512, 1024x1024

Measures CPU vs GPU timing and speedup ratio.
"""

# CRITICAL: Set x64 BEFORE any jax.numpy imports
import jax
jax.config.update("jax_enable_x64", True)

import numpy as np
import time
import json
from pathlib import Path
import jax.numpy as jnp

# Import benchmark utilities
from benchmark_utils import (
    setup_benchmark, check_finite_tree, compute_stats,
    add_benchmark_args, DEFAULT_N_REPS
)

# Parse CLI arguments
parser = add_benchmark_args()
args = parser.parse_args()
N_REPS = args.n_reps
# Use None for backend check since this is a mixed CPU/GPU benchmark
EXPECTED_BACKEND = None

# Configuration
GRID_SIZES = [64, 128, 256, 512, 1024]
N_STEPS = 1000
D = 0.01
N_WARMUP = 2

print("Scaling Study Benchmark")
print("=" * 60)
device_str = setup_benchmark(expected_backend=EXPECTED_BACKEND)
print(f"Grid sizes: {GRID_SIZES}")
print(f"Steps per run: {N_STEPS}")
print(f"N_REPS: {N_REPS}")
print("=" * 60)

results = {'grids': []}

for N in GRID_SIZES:
    print(f"\n--- Grid: {N}x{N} ({N*N:,} DOFs) ---")

    dx = 1.0 / N
    dt = 0.0001

    # Setup wavenumbers
    kx = np.fft.fftfreq(N, dx) * 2 * np.pi
    ky = np.fft.fftfreq(N, dx) * 2 * np.pi
    KX, KY = np.meshgrid(kx, ky, indexing='ij')
    lap_eig = -(KX**2 + KY**2)

    # Initial condition
    u0 = np.random.randn(N, N)

    # =========================================================================
    # CPU (NumPy)
    # =========================================================================
    def numpy_solve(u0, n_steps):
        u = u0.copy()
        for _ in range(n_steps):
            u_hat = np.fft.fft2(u)
            u_hat = u_hat * (1 + 0.5*dt*D*lap_eig) / (1 - 0.5*dt*D*lap_eig)
            u = np.real(np.fft.ifft2(u_hat))
        return u

    # Warmup
    for _ in range(N_WARMUP):
        _ = numpy_solve(u0, 10)

    # Time
    times_cpu = []
    for _ in range(N_REPS):
        start = time.perf_counter()
        result_cpu = numpy_solve(u0, N_STEPS)
        times_cpu.append(time.perf_counter() - start)

    # Check for NaN/Inf
    if not np.all(np.isfinite(result_cpu)):
        raise ValueError(f"CPU result contains NaN/Inf at N={N}")

    cpu_median, cpu_iqr = compute_stats(times_cpu)
    print(f"  CPU: {cpu_median:.2f} +/- {cpu_iqr:.2f} s")

    # =========================================================================
    # GPU (JAX)
    # =========================================================================
    u0_jax = jnp.array(u0)
    lap_eig_jax = jnp.array(lap_eig)

    @jax.jit
    def jax_step(u, lap_eig):
        u_hat = jnp.fft.fft2(u)
        u_hat = u_hat * (1 + 0.5*dt*D*lap_eig) / (1 - 0.5*dt*D*lap_eig)
        return jnp.real(jnp.fft.ifft2(u_hat))

    def jax_solve(u0, n_steps, lap_eig):
        u = u0
        for _ in range(n_steps):
            u = jax_step(u, lap_eig)
        return u

    # Warmup (compile)
    for _ in range(N_WARMUP + 1):
        _ = jax_solve(u0_jax, 10, lap_eig_jax).block_until_ready()

    # Time
    times_gpu = []
    for _ in range(N_REPS):
        start = time.perf_counter()
        result_gpu = jax_solve(u0_jax, N_STEPS, lap_eig_jax)
        result_gpu.block_until_ready()
        times_gpu.append(time.perf_counter() - start)

    # Check for NaN/Inf
    check_finite_tree(result_gpu, f"GPU result N={N}")

    gpu_median, gpu_iqr = compute_stats(times_gpu)

    speedup = cpu_median / gpu_median
    print(f"  GPU: {gpu_median:.2f} +/- {gpu_iqr:.2f} s")
    print(f"  Speedup: {speedup:.1f}x")

    results['grids'].append({
        'N': N,
        'DOFs': N * N,
        'cpu_median_s': float(cpu_median),
        'cpu_iqr_s': float(cpu_iqr),
        'gpu_median_s': float(gpu_median),
        'gpu_iqr_s': float(gpu_iqr),
        'speedup': float(speedup),
    })

# Save results
results['config'] = {
    'n_steps': N_STEPS,
    'n_reps': N_REPS,
    'D': D,
    'device': device_str,
    'dtype': 'float64',
}

output_path = Path(__file__).parent / 'results' / 'scaling.json'
with open(output_path, 'w') as f:
    json.dump(results, f, indent=2)
print(f"\nResults saved to {output_path}")

# =============================================================================
# Summary Table
# =============================================================================
print("\n" + "=" * 70)
print("SUMMARY (Table 9)")
print("=" * 70)
print(f"{'Grid':>12} {'DOFs':>12} {'CPU (s)':>12} {'GPU (s)':>12} {'Speedup':>10}")
print("-" * 70)
for g in results['grids']:
    print(f"{g['N']:>5}x{g['N']:<5} {g['DOFs']:>12,} {g['cpu_median_s']:>12.2f} {g['gpu_median_s']:>12.2f} {g['speedup']:>10.1f}x")
print("=" * 70)

# =============================================================================
# Generate Figure
# =============================================================================
import matplotlib.pyplot as plt

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))

dofs = [g['DOFs'] for g in results['grids']]
cpu_times = [g['cpu_median_s'] for g in results['grids']]
gpu_times = [g['gpu_median_s'] for g in results['grids']]
speedups = [g['speedup'] for g in results['grids']]

# Left: CPU vs GPU timing (log-log)
ax1.loglog(dofs, cpu_times, 'o-', label='CPU (NumPy)', color='#1f77b4', linewidth=2, markersize=8)
ax1.loglog(dofs, gpu_times, 's-', label='GPU (JAX)', color='#2ca02c', linewidth=2, markersize=8)
ax1.set_xlabel('Degrees of Freedom', fontsize=12)
ax1.set_ylabel('Time (seconds)', fontsize=12)
ax1.set_title('Scaling: CPU vs GPU', fontsize=14)
ax1.legend(fontsize=11)
ax1.grid(True, alpha=0.3)

# Mark crossover region
ax1.axvline(x=10000, color='gray', linestyle='--', alpha=0.5)
ax1.text(10000, max(cpu_times)*0.5, 'Crossover\n~100x100', ha='center', fontsize=9, color='gray')

# Right: Speedup vs DOFs
ax2.semilogx(dofs, speedups, 'o-', color='#d62728', linewidth=2, markersize=8)
ax2.set_xlabel('Degrees of Freedom', fontsize=12)
ax2.set_ylabel('GPU Speedup (x)', fontsize=12)
ax2.set_title('GPU Speedup vs Problem Size', fontsize=14)
ax2.grid(True, alpha=0.3)
ax2.axhline(y=1, color='gray', linestyle='--', alpha=0.5)

# Add value labels
for x, y in zip(dofs, speedups):
    ax2.annotate(f'{y:.1f}x', (x, y), textcoords="offset points",
                xytext=(0, 10), ha='center', fontsize=9)

plt.tight_layout()
fig_path = Path(__file__).parent.parent / 'figures' / 'scaling.png'
plt.savefig(fig_path, dpi=150, bbox_inches='tight')
print(f"Figure saved to {fig_path}")
plt.close()
